internal/contributebot/design/merge.md

```diff
@@ -0,0 +1,101 @@
+# Contribute Bot Auto-Merge Design
+
+Author: [@zombiezen][]<br>
+Date: 2018-12-17<br>
+Part of [#687][]
+
+[@zombiezen]: https://github.com/zombiezen
+[#687]: https://github.com/google/go-cloud/issues/687
+
+## Objective
+
+As a maintainer of Go Cloud, it is desirable to have pull requests merged in as
+quickly as they can be merged after receiving approvals, assuming there are no
+test breakages or merge conflicts. Currently, merging requires a manual process:
+
+```
+while branch is not up-to-date with master {
+  press update branch button
+  wait for Travis run to finish
+}
+click squash-and-merge
+```
+
+This loop is especially expensive the longer the waiting is, since it increases
+the likelihood that the branch is not up-to-date and thus causing the cycle to
+increase iterations (more waiting!).
+
+The objective of this design is to reduce the amount of maintainer [toil][] in
+this process.
+
+[toil]: https://landing.google.com/sre/sre-book/chapters/eliminating-toil/
+
+## Overview
+
+To reduce maintainer toil, this document proposes a new set of interactions for
+Contribute Bot. If a user writes a comment on a pull request that starts with
+the text "/merge", Contribute Bot will:
+
+1.  Verify that the commenter has `write` or `admin` access to the repository.
+    If they do not, do not proceed and leave an error comment on the PR.
+2.  Verify that the pull request has the approvals required by the [GitHub
+    branch protections][required reviews]. If it does not, do not proceed and
+    leave an error comment on the PR.
+3.  Verify that there are no other commits on the pull request branch since the
+    last approved commit (or a manually entered commit hash entered by the
+    commenter) other than those created by Contribute Bot. If such commits
+    exist, then do not proceed and leave a comment on the PR mentioning that
+    "/merge COMMITHASH" will bypass this. While this is stricter than what the
+    branch protections might allow, it avoids accidentally merging in unreviewed
+    changes.
+3.  Leave a comment on the PR to the effect that Contribute Bot has received the
+    request and that it will report back if it is unable to merge the PR.
+4.  If the pull request is not up to date with the changes on the base branch,
+    then create a merge commit on the pull request branch. If the merge commit
+    cannot be created due to merge conflicts, then stop trying to merge the PR
+    and make an error comment on the PR.
+5.  Wait for any necessary check runs to report completion. If any report
+    failure, then stop trying to merge the PR and make an error comment on the
+    PR.
+6.  Merge the pull request using the project's allowed merge method (`squash`
```

Comment:
The problem here is that the squashed commit message contains all the commits' messages, which may be things like "addressed reviewer comments" or "added test". (At least that's what mine look like.) I've been manually editing them at this point.

Reply:
Right, this isn't pulling from the commit messages, but the PR title and first comment body.

An alternative here could be to have the rest of the `/merge` comment be the commit message, but that seems onerous to me in the common case where we assume that the contributor wrote a good PR message and/or the reviewer edited the PR message to be good.

Reply:
Hmm... What if the first comment body is "WIP: DO NOT MERGE" but later that goes away?

Or if stuff was added during review and that should be in the comment?

I guess we could edit the first comment as you suggest, but that feels like rewriting history.

Or the comment after "/merge" could become the commit comment.

Anyway, I don't want to hold up the whole thing on this.

Reply:
I share the concern about rewriting history: your comments echo the same back-and-forth I had internally before sending out the doc. The thing with using the "/merge" body is that it introduces burden on the maintainer to curate the commit message, which I see as more the responsibility of the PR author. GitHub does allow watching the history of a PR comment, so hopefully that alleviates the concern.

We can experiment and find out what flow works best.